### PR TITLE
ENH: Expose c1 and c2 of scalar_search_wolfe to fmin_bfgs

### PR DESCRIPTION
**Pull request by TLeitzbach**
**What does this implement/fix?**

This PR exposes the c1 and c2 params of the scalar_search_wolfe1 / 2 in optimize._linesearch.
These parameters cannot currently be modified by calling fmin_bfgs or _minimize_bfgs.

Check list:
- I ran the optimize tests locally 
- I build the docs locally
- I added the corresponding doc string parameters

### CODE REVIEW

File: scipy/optimize/_optimize.py [hunk 1]
```diff
@@ -1375,7 +1381,7 @@ def fmin_bfgs(f, x0, fprime=None, args=(), gtol=1e-5, norm=np.inf,
 def _minimize_bfgs(fun, x0, args=(), jac=None, callback=None,
                    gtol=1e-5, norm=np.inf, eps=_epsilon, maxiter=None,
                    disp=False, return_all=False, finite_diff_rel_step=None,
-                   xrtol=0, **unknown_options):
+                   xrtol=0, c1=1e-4, c2=.9, **unknown_options):
```
Comment (dschmitz89) [inline scipy/optimize/_optimize.py:1384]:
```suggestion
                   xrtol=0, c1=1e-4, c2=0.9, **unknown_options):
```

For readability.

**Review by dschmitz89 — CHANGES_REQUESTED** [review]
Thanks for the PR @TLeitzbach!

For optimization experts, this can be a useful feature, so am I for adding this functionality.

Currently, this change is missing tests. We need tests that changing the parameters actually changes the results. Examples could be to check that the number of iterations becomes higher with a tighter convergence tolerance but I am not really a domain expert. Does that sound sensible?

**Comment by TLeitzbach** [comment]
@dschmitz89 thank you!

You are right, I added two tests using the Rosenbrock function.
One that illustrates the impact on the number of iterations when altering c2
and another that shows that an insufficiently low value of c1 in combination with bad starting parameters fails to converge due to precision loss.

File: scipy/optimize/_optimize.py [hunk 2]
```diff
@@ -1244,7 +1244,7 @@ def _line_search_wolfe12(f, fprime, xk, pk, gfk, old_fval, old_old_fval,
 
 def fmin_bfgs(f, x0, fprime=None, args=(), gtol=1e-5, norm=np.inf,
               epsilon=_epsilon, maxiter=None, full_output=0, disp=1,
-              retall=0, callback=None, xrtol=0):
+              retall=0, callback=None, xrtol=0, c1=1e-4, c2=.9):
```
Comment (dschmitz89) [inline scipy/optimize/_optimize.py:1247]:
```suggestion
              retall=0, callback=None, xrtol=0, c1=1e-4, c2=0.9):
```

File: scipy/optimize/_optimize.py [hunk 3]
```diff
@@ -1281,6 +1281,10 @@ def fmin_bfgs(f, x0, fprime=None, args=(), gtol=1e-5, norm=np.inf,
         Relative tolerance for `x`. Terminate successfully if step
         size is less than ``xk * xrtol`` where ``xk`` is the current
         parameter vector.
+    c1 : float, default: 1e-4
+        Parameter for Armijo condition rule.
+    c2 : float, default: .9
```
Comment (dschmitz89) [inline scipy/optimize/_optimize.py:1286]:
```suggestion
    c2 : float, default: 0.9
```

File: scipy/optimize/tests/test_optimize.py [hunk 4]
```diff
@@ -237,6 +237,26 @@ def test_bfgs_xrtol(self):
...
+        x0 = [10.3, 20.7, 10.8, 1.9, -1.2]
+        res = optimize.minimize(optimize.rosen,
+                                x0, method='bfgs', options={'c1': 1e-8})
+        ref = optimize.minimize(optimize.rosen,
+                                x0, method='bfgs', options={'c1': 1e-1})
+
+        assert res.nfev > ref.nfev
+
+    def test_bfgs_c2(self):
+        # test for #18977 test number of iterations for modification of c2 parameter
```
Comment (dschmitz89) [inline scipy/optimize/tests/test_optimize.py:252]:
```suggestion
        # test that modification of c2 parameter results in different number of iterations
```

File: scipy/optimize/tests/test_optimize.py [hunk 5]
```diff
@@ -237,6 +237,26 @@ def test_bfgs_xrtol(self):
...
+        x0 = [1.3, 0.7, 0.8, 1.9, 1.2]
+        res = optimize.minimize(optimize.rosen,
+                                x0, method='bfgs', options={'c2': .9})
+        ref = optimize.minimize(optimize.rosen,
+                                x0, method='bfgs', options={'c2': 1e-2})
+        assert res.nit > ref.nit
```
Comment (dschmitz89) [inline scipy/optimize/tests/test_optimize.py:258]:
```suggestion
        res_default = optimize.minimize(optimize.rosen,
                                        x0, method='bfgs', options={'c2': .9})
        res_mod = optimize.minimize(optimize.rosen,
                                    x0, method='bfgs', options={'c2': 1e-2})
        assert res_default.nit > res_mod.nit
```

File: scipy/optimize/tests/test_optimize.py [hunk 6]
```diff
@@ -237,6 +237,26 @@ def test_bfgs_xrtol(self):
...
+    def test_bfgs_c1(self):
+        # test for #18977 insufficiently low value of c1 leads to precision loss
+        # for poor starting parameters
+        x0 = [10.3, 20.7, 10.8, 1.9, -1.2]
+        res = optimize.minimize(optimize.rosen,
+                                x0, method='bfgs', options={'c1': 1e-8})
+        ref = optimize.minimize(optimize.rosen,
+                                x0, method='bfgs', options={'c1': 1e-1})
+
+        assert res.nfev > ref.nfev
```
Comment (dschmitz89) [inline scipy/optimize/tests/test_optimize.py:249]:
```suggestion
        res_c1_small = optimize.minimize(optimize.rosen,
                                         x0, method='bfgs', options={'c1': 1e-8})
        res_c1_big = optimize.minimize(optimize.rosen,
                                       x0, method='bfgs', options={'c1': 1e-1})

        assert res_c1_small.nfev > res_c1_big.nfev
```

**Review by dschmitz89 — APPROVED** [review]
Thanks @TLeitzbach for the updates. The tests make sense. My suggestions are mostly to enhance readability and maintainability in future.

**Comment by TLeitzbach** [comment]
Thanks @dschmitz89 for the suggestions and the feedback!
I have committed them as suggested.

File: scipy/optimize/_optimize.py [hunk 7]
```diff
@@ -1443,7 +1458,7 @@ def _minimize_bfgs(fun, x0, args=(), jac=None, callback=None,
         try:
             alpha_k, fc, gc, old_fval, old_old_fval, gfkp1 = \
                      _line_search_wolfe12(f, myfprime, xk, pk, gfk,
-                                          old_fval, old_old_fval, amin=1e-100, amax=1e100)
+                                          old_fval, old_old_fval, amin=1e-100, amax=1e100, **kwargs)
```
Comment (j-bowhay) [inline scipy/optimize/_optimize.py:1461]:
```suggestion
                                          old_fval, old_old_fval, amin=1e-100,
                                          amax=1e100, c1=c1, c2=c2)
```

File: scipy/optimize/_optimize.py [hunk 8]
```diff
@@ -1406,8 +1412,17 @@ def _minimize_bfgs(fun, x0, args=(), jac=None, callback=None,
...
+        Parameter for Armijo condition rule.
+    c2 : float, default: .9
+        Parameter for curvature condition rule.
+
     """
     _check_unknown_options(unknown_options)
+    kwargs = {
+        'c1': c1,
+        'c2': c2,
+    }
```
Comment (j-bowhay) [inline scipy/optimize/_optimize.py:1425]:
```suggestion
```

**Review by j-bowhay — CHANGES_REQUESTED** [review]
Thanks @TLeitzbach! Here we can just pass c1 & c2 as keyword arguments rather than creating a kwargs dict

**Comment by j-bowhay** [comment]
Also, fmin_bfgs is a legacy function so I am not sure we actually want to make enhancements. Can you not expose it through `scipy.optimize.minimize` instead?

**Comment by TLeitzbach** [comment]
@j-bowhay cheers for your feedback! 
I fully agree and committed them as suggested.

**Review by dschmitz89 — APPROVED** [review]

**Comment by TLeitzbach** [comment]
@j-bowhay could you please resolve the changes requested flag if you are satisfied?

**Comment by TLeitzbach** [comment]
Hi @andyfaff / @j-bowhay is there anything I can do to help you close this?

**Comment by j-bowhay** [comment]
I will leave the final call on this to @andyfaff/another optimize regular

File: scipy/optimize/_optimize.py [hunk 9]
```diff
@@ -1413,6 +1419,11 @@ def _minimize_bfgs(fun, x0, args=(), jac=None, callback=None,
     xrtol : float, default: 0
         Relative tolerance for `x`. Terminate successfully if step size is
         less than ``xk * xrtol`` where ``xk`` is the current parameter vector.
+    c1 : float, default: 1e-4
```
Comment (andyfaff) [inline scipy/optimize/_optimize.py:1422]: Are there guidelines for what the values of c1 and c2 should be? e.g. non-negative, in a range ...?

**Comment by andyfaff** [comment]
I'm happy for this to be merged, but I would like to have guidelines on how to choose parameter values.
Can you also squash the commits in this PR as well please?

File: scipy/optimize/_optimize.py (hunk 9 above)
Comment (paulestano) [reply scipy/optimize/_optimize.py:1422]: These are the typical values used in [Numerical Optimization from Nocedal (p.54)](https://link.springer.com/book/10.1007/978-0-387-40065-5) and in [L-BFGS paper](https://link.springer.com/article/10.1007/BF01589116). They also must be such that $0 < c2 < c1 < 1$ this should potentially be check before running the line-search.

**Pull request merged**